newrelic/newrelic-diagnostics-cli
# Validate nrjmx mbeans

# Issue
Not an issue, we are adding new features that I'll describe in the next section. 
# Goals
1)Adding a task Infra/Env/NrjmxBeans: Validate that the mbeans that the customer wants to query with the JMX integration are available for the integration to pull from.
2)Add two new fields to the struct JmxConfig found in the task Infra/Config/ValidateJMX: This task is mainly used by TSEs to troubleshoot that a JMX server cannot connect to our JMX integration, and not so much for a customer to understand WHY a connection is failing. This is because we only output the error found, but we do not judge what the error means.
With that same purpose of using this task only to provide data to TSEs to troubleshoot, I have added two new fields to this task struct that contains information about the Java version and the cmdline arguments for JMX. This specific information was requested by infra TSEs after I met with them and they said they use those 2 pieces of information to troubleshoot alongside the error that nrdiag provides. This information will only be provided to TSEs and not customers as it only shows up in the payload result, and not in the summary result.

# Implementation Details

1)Infra/Env/NrjmxMbeans:
-It will parse through the JMX integration metrics files listed under in collection_files field found in the jmx-config.yml
-It will grab the all queries defined in those metrics files and run them against the command `{queryname} | nrjmx -H %s -P %s -v -d -`
-If any of the queries return an empty object, it will warn the customer about this query not being available to the integration

2)Infra/Config/ValidateJMX:
-To get data the java version, I am leveraging the Java/Env/Version task to get the payload string
-To get the cmdline arguments for JMX, I am running this command `ps -ef | grep jmx` and converting the output []bytes to a string. Then adding that string to our result payload for TSE to interpret that string to the best of their knowledge. 

# How to Test
I have added a couple of unit test to the new task: Infra/Env/NrjmxMbeans.
I have also smoke tested in a linux environment that has JVM processes running and one of them had JMX enabled. I installed and configured the JMX integration to point to the incorrect port for the JMX server just to see the task fail with a very common mistake customers make. 

If you are interested in smoke testing this task, you'll have to follow these instructions to set the JMX integration: https://docs.newrelic.com/docs/integrations/host-integrations/host-integrations-list/jmx-monitoring-integration

To enable JMX metrics on your testing app, you can pass arguments like these when you start up your java app: 
java-jar -Dcom.sun.management.jmxremote.password.file=jmxremote.password-Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.authenticate=true-Dcom.sun.management.jmxremote.port=9010-Dcom.sun.management.jmxremote.ssl=false -pathtoyourapp.jar

Then you'll have to checkout into this branch and run ./scripts/build.sh to get the linux binary and drop the binary in your JMX server. This task does not run yet on windows. Though I am planning to on my next PR to fix a bug in ValidateJMX task that does not allow for this task to run on Windows.
